pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -444,6 +445,268 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
+        // proportionately to 32/8, to fit within the remaining 40%
+        //
+        //uint256 weightSumMultiplier = FixedPoint.ONE.divDown(FixedPoint.ONE - normalizedWeight);
+        weightSumAfterAdd = weightSumBeforeAdd.mulUp(FixedPoint.ONE.divDown(FixedPoint.ONE - normalizedWeight));
+
+        // Now we need to make sure the other token weights don't get scaled down below the minimum
+        // normalized weight[i] = denormalized weight[i] / weightSumAfterAdd
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        for (uint256 i = 0; i < tokens.length; i++) {
+            _require(
+                _getTokenData(tokens[i]).decodeUint64(_END_DENORM_WEIGHT_OFFSET).uncompress64(_MAX_DENORM_WEIGHT).divUp(
+                    weightSumAfterAdd
+                ) >= WeightedMath._MIN_WEIGHT,
+                Errors.MIN_WEIGHT
+            );
+        }
+
+        // Calculate the bptAmountOut equivalent to adding the token at the given weight
+        // The BPT price of a token = S / (B/Wn) = (S * Wn) / B;
+        // where S = totalSupply, Wn = normalized weight, and B = balance
+        //
+        // Since the BPT prices of existing tokens should stay constant when adding a token,
+        // Let Sb, Wnb = total supply and normalized weight before the add; and
+        //     Sa, Wna = total supply and normalized weight after the add
+        // Then: (Sa * Wna) / B = (Sb * Wnb) / B
+        //       Sa * Wna = Sb * Wnb
+        //       Sa = Sb * (Wnb / Wna)
+        // Since the normalized weights (Wn) = denormalized weights (Wd) / weightSum (WS), we have:
+        // Sa = Sb * (Wd / WSb)  - denormalized weights don't change, so there is just one Wd
+        //            ---------
+        //           (Wd / WSa)
+        // Sa = Sb * WSa / WSb = totalSupply after the add
+        //
+        // Then the bptAmountOut is the delta in the total supply:
+        // bptAmountOut = Sa - Sb
+        //              = Sb * WSa / WSb - Sb
+        //              = Sb * (WSa / WSb - 1)
+        //
+        // In our example, Sa is the totalSupply on initialization of the pool. If the balances are 400 and 0.5:
+        // Sb = (400^0.8) * (0.5^0.2) * 2 = 210.1222
+        // Sa = 210.1222 * (2.5 / 1.0) = 525.3056
+        // bptAmountOut = 210.1222 * (2.5 / 1.0 - 1) = 315.1833
+        // The added token should represent 60% of the total supply, after the add, and in fact:
+        // 315.1833 / 525.3056 = ~ 0.6
+        //
+        uint256 weightSumRatio = weightSumAfterAdd.divDown(weightSumBeforeAdd);
+        bptAmountOut = totalSupply().mulDown(weightSumRatio.sub(FixedPoint.ONE));
+
+        // Validate that the actual BPT price
+        uint256 actualBptPrice = totalSupply().mulDown(weightSumRatio).mulDown(normalizedWeight).divUp(
+            _upscale(tokenAmountIn, _computeScalingFactor(token))
+        );
+
+        _require(actualBptPrice >= minBptPrice, Errors.MIN_BPT_PRICE_ADD_TOKEN);
```

Comment:
I'm unsure about this whole thing. Should we receive a `minBPTPrice`? Is that the equivalent to `minTokensOut` in `removeToken`?

If so, it seems this belongs in WeightedMath, not here.